Network/Socket/ByteString/IO.hsc

```diff
@@ -248,3 +266,87 @@ withIOVec cs f =
         unsafeUseAsCStringLen s $ \(sPtr, sLen) ->
         poke ptr $ IOVec sPtr (fromIntegral sLen)
 #endif
+
+-- | Send data from the socket using sendmsg(2).
+sendMsg :: Socket       -- ^ Socket
+        -> SockAddr     -- ^ Destination address
+        -> [ByteString] -- ^ Data to be sent
+        -> [Cmsg]       -- ^ Control messages
+        -> [MsgFlag]    -- ^ Message flags
+        -> IO Int       -- ^ The length actually sent
+sendMsg _ _    []  _ _ = return 0
+sendMsg s addr bss cmsgs flags = do
+  sz <- withSockAddr addr $ \addrPtr addrSize ->
+    withIOVec bss $ \(iovsPtr, iovsLen) -> do
+      withCmsgs cmsgs $ \ctrlPtr ctrlLen -> do
+        let msgHdr = MsgHdr {
+                msgName    = addrPtr
+              , msgNameLen = fromIntegral addrSize
+              , msgIov     = iovsPtr
+              , msgIovLen  = fromIntegral iovsLen
+              , msgCtrl    = castPtr ctrlPtr
+              , msgCtrlLen = fromIntegral ctrlLen
+              , msgFlags   = 0
+              }
+            cflags = packBits msgFlagMapping flags
+        withFdSocket s $ \fd ->
+          with msgHdr $ \msgHdrPtr ->
+            throwSocketErrorWaitWrite s "Network.Socket.ByteString.sendMsg" $
+              c_sendmsg fd msgHdrPtr cflags
+  return $ fromIntegral sz
+
+-- | Receive data from the socket using recvmsg(2).
+--   The receive buffers are created according to the second argument.
+--   If the length of received data is less than the total of
+--   the second argument, the buffers are truncated properly.
+--   So, only the received data can be seen.
+recvMsg :: Socket    -- ^ Socket
+        -> [Int]     -- ^ A list of length of data to be received
+                     --   If the total length is not large enough,
+                     --   'MSG_TRUNC' is returned
+        -> Int       -- ^ The buffer size for control messages.
+                     --   If the length is not large enough,
+                     --   'MSG_CTRUNC' is returned
+        -> [MsgFlag] -- ^ Message flags
+        -> IO (SockAddr, [ByteString], [Cmsg], [MsgFlag]) -- ^ Source address, received data, control messages and message flags
```

Comment:
> Like `recvBuf` vs. `recv` I'd like an API that would write to pre-allocated buffers rather than creating the lists of bytestrings, cmsg, and flags, for lower overhead.

This being Haskell, the primary API should be a safe one, that does not mutate memory. If there's a compelling need for an unsafe API that writes to mutable buffers, perhaps that can be added, but my take is that a user who wants that, can do whatever they were going to do with that mutable memory also in C, and expose the composite functionality via the FFI.

Reply:
@maoe Yes, I'm planning to provide `recvMsgBuf` and implement `recvMsg` atop it.

Reply:
`sendBufMsg` and `recvBufMsg` have been implemented.